pkg/relayer/miner/miner.go

```diff
@@ -0,0 +1,408 @@
+	return filter.EitherSuccess(ctx, eitherClaimedSessions)
+}
+
+// submitProofs maps over the given claimedSessions observable. For each session,
+// it calculates and waits for the earliest block height at which it is safe to
+// submit a proof and does so. It then maps any errors to a new observable which
+// are subsequently logged. It does not block as map operations run in their own
+// goroutines.
+func (mnr *miner) submitProofs(
+	ctx context.Context,
+	claimedSessions observable.Observable[relayer.SessionTree],
+) {
+	// relayer.SessionTree ==> relayer.SessionTree
+	sessionsWithOpenProofWindow := channel.Map(ctx, claimedSessions, mnr.mapWaitForOpenProofWindow)
+
+	failedSubmitProofSessions, failedSubmitProveSessionsPublishCh := channel.NewObservable[relayer.SessionTree]()
+
+	// relayer.SessionTree ==> relayer.SessionTree
+	eitherProvenSessions := channel.Map(
+		ctx, sessionsWithOpenProofWindow,
+		mnr.newMapProveSessionFn(failedSubmitProveSessionsPublishCh),
+	)
+
+	// TODO_TECHDEBT: pass failed submit proof sessions to some retry mechanism.
+	_ = failedSubmitProofSessions
+	logging.LogErrors(ctx, filter.EitherError(ctx, eitherProvenSessions))
+}
+
+// validateConfigAndSetDefaults ensures that the miner has been configured with
+// a hasher and uses the default hasher if not.
+func (mnr *miner) validateConfigAndSetDefaults() error {
+	if mnr.hasher == nil {
+		mnr.hasher = defaultHasher
+	}
+	return nil
+}
+
+// mineRelays maps over the servedRelays observable, applyging the mapMineRelay
+// method to each relay. It returns an observable of the mined relays.
+func (mnr *miner) mineRelays(
+	ctx context.Context,
+	servedRelays observable.Observable[servicetypes.Relay],
+) observable.Observable[either.Either[minedRelay]] {
+	// servicetypes.Relay ==> either.Either[minedRelay]
+	return channel.Map(ctx, servedRelays, mnr.mapMineRelay)
+}
+
+// mapMineRelay is intended to be used as a MapFn. It hashes the relay and compares
+// its difficulty to the minimum threshold. If the relay difficulty is sifficient,
+// it returns an either populated with the minedRelay value. Otherwise, it skips
+// the relay. If it encounters an error, it returns an either populated with the
+// error.
+func (mnr *miner) mapMineRelay(
+	_ context.Context,
+	relay servicetypes.Relay,
+) (_ either.Either[minedRelay], skip bool) {
+	relayBz, err := relay.Marshal()
+	if err != nil {
+		return either.Error[minedRelay](err), true
+	}
+
+	// Is it correct that we need to hash the key while smst.Update() could do it
+	// since smst has a reference to the hasher
+	mnr.hasher.Write(relayBz)
+	relayHash := mnr.hasher.Sum(nil)
+	mnr.hasher.Reset()
+
+	if !protocol.BytesDifficultyGreaterThan(relayHash, defaultRelayDifficulty) {
+		return either.Success(minedRelay{}), true
+	}
+
+	return either.Success(minedRelay{
+		Relay: relay,
+		bytes: relayBz,
+		hash:  relayHash,
+	}), false
+}
+
+// addReplayToSessionTree maps over the eitherMinedRelays observable, applying the
```

Comment:
_tl;dr I think we shold remove this function altogether._

- I was thinking of having this function take in a function as a param (but realized it's overkill)
- This seems like an unnecessary wrapper (having it in place would add clarity)
- All we're doing is mapping things to errors but the name was very confusing.